src/Symfony/Component/HttpClient/Rfc9111HttpClient.php

```diff
@@ -0,0 +1,616 @@
+            return $this->asyncStream($asyncResponses, $timeout);
+        }
+
+        if (!$asyncResponses && !$clientResponses) {
+            return new ResponseStream(MockResponse::stream($mockResponses, $timeout));
+        }
+
+        if (!$asyncResponses && !$mockResponses) {
+            return $this->client->stream($clientResponses, $timeout);
+        }
+
+        return new ResponseStream((function () use ($mockResponses, $asyncResponses, $clientResponses, $timeout) {
+            yield from MockResponse::stream($mockResponses, $timeout);
+            yield $this->client->stream($clientResponses, $timeout);
+            yield $this->asyncStream($asyncResponses, $timeout);
+        })());
+    }
+
+    /**
+     * Returns a hash representing the request details.
+     *
+     * @param string $method the request method
+     * @param string $url    the request URL
+     *
+     * @return string the request hash
+     */
+    private function getRequestHash(string $method, string $url): string
+    {
+        return hash('xxh3', $method.$url);
+    }
+
+    /**
+     * Generates a unique metadata key based on the request hash and varying headers.
+     *
+     * @param string $requestHash       a hash representing the request details
+     * @param array  $normalizedHeaders normalized headers of the request
+     * @param array  $varyFields        headers to consider for building the variant key
+     *
+     * @return string the metadata key composed of the request hash and variant key
+     */
+    private function getMetadataKey(string $requestHash, array $normalizedHeaders, array $varyFields): string
+    {
+        $variantKey = hash(
+            'xxh3',
+            [] === $varyFields
+                ? 'NO-VARY'
+                : $this->buildVariantKey($normalizedHeaders, $varyFields)
+        );
+
+        return "metadata_{$requestHash}_{$variantKey}";
+    }
+
+    /**
+     * Build a variant key for caching, given an array of normalized headers and the vary fields.
+     *
+     * The key is a pipe-separated string of "header=value" pairs, with the special case of "header=" for headers that are not present.
+     *
+     * @param array $normalizedHeaders normalized headers
+     * @param array $varyFields        vary fields
+     *
+     * @return string the variant key
+     */
+    private function buildVariantKey(array $normalizedHeaders, array $varyFields): string
+    {
+        $parts = [];
+        foreach ($varyFields as $field) {
+            $lower = strtolower($field);
+            if (!isset($normalizedHeaders[$lower])) {
+                $parts[] = $field.'=';
+            } else {
+                $joined = \is_array($normalizedHeaders[$lower])
+                    ? implode(',', $normalizedHeaders[$lower])
+                    : (string) $normalizedHeaders[$lower];
+                $parts[] = $field.'='.$joined;
+            }
+        }
+
+        return implode('|', $parts);
```

Comment:
what if we have header values containing a `|` ? Currently, this would allow matching a different key.

Reply:
Adressed.